File: machine/cpp/core/rom_boot_manager.cpp
#include "core/rom_boot_manager.h"
#include "core/system.h"
#include "rompack/loader.h"

namespace bmsx {

std::unique_ptr<RomBootPlan> RomBootManager::buildBootPlan(
	const u8* systemRom, size_t systemSize,
	const u8* cartridge, size_t cartSize)
{
	(void)cartridge;
	(void)cartSize;
	auto plan = std::make_unique<RomBootPlan>();

	loadSystemRomPackageFromRom(systemRom, systemSize, plan->systemLayer, nullptr, "system");
	plan->systemLayer.machine = defaultSystemMachineManifest();
	plan->systemLayer.entryPoint = systemBootEntryPath();

	return plan;
}

} // namespace bmsx

File: machine/ts/core/rom_boot_manager.ts
import {
	buildRuntimeRomLayer,
	buildSystemRuntimeRomLayer,
	type RuntimeRomLayer,
} from '../rompack/loader';
import { SYSTEM_BOOT_ENTRY_PATH, SYSTEM_MACHINE_MANIFEST } from './system';

export type RomBootPlan = {
	systemLayer: RuntimeRomLayer;
	cartLayer: RuntimeRomLayer | null;
};

export class RomBootManager {
	public async buildBootPlan(options: { systemRom: Uint8Array; cartridge?: Uint8Array }): Promise<RomBootPlan> {
		const systemLayer = await buildSystemRuntimeRomLayer({
			blob: options.systemRom,
			machine: SYSTEM_MACHINE_MANIFEST,
			entry_path: SYSTEM_BOOT_ENTRY_PATH,
		});
		const cartLayer = options.cartridge
			? await buildRuntimeRomLayer({ blob: options.cartridge, id: 'cart' })
			: null;

		return {
			systemLayer,
			cartLayer,
		};
	}
}
